docs/future/compatibility.md

```diff
@@ -0,0 +1,174 @@
+
+2. Additional libraries can provide compatible adapters
+
+React API-compatible libraries such as `preact` and `inferno` would be able to provide adapters to Enzyme
+for their corresponding libraries, and be able to take full advantage of Enzyme's APIs.
+
+3. Better user experience (ie, bundlers won't complain about missing deps)
+
+Enzyme has had a long-standing issue with static-analysis bundlers such as Webpack and Browserify because
+of our usage of internal React APIs. With this change, this would be minimized if not removed entirely,
+since these things can be localized into the adapter modules, and users will only install the ones they need.
+
+Additionally, we can even attempt to remove the use of internal react APIs by lobbying for react-maintained packages
+such as `react-test-renderer` to utilize the React Standard Tree (RST) format (details below).
+
+4. Standardization and interopability with other tools
+
+If we can agree on the tree format (specified below as "React Standard Tree"), other tools can start to use and
+understand this format as well. Standardization is a good thing, and could allow tools to be built that maybe
+don't even exist yet.
+
+
+## Proposal
+
+
+### React Standard Tree (RST)
+
+This proposal hinges on a standard tree specification. Keep in mind that this tree needs to account for more
+than what is currently satisfied by the output of something like `react-test-renderer`, which is currently
+only outputting the "host" nodes (ie, HTML elements). We need a tree format that allows for expressing a full
+react component tree, including composite components.
+
+```js
+// Strings and Numbers are rendered as literals.
+type LiteralValue = string | number
+
+// A "node" in an RST is either a LiteralValue, or an RSTNode
+type Node = LiteralValue | RSTNode
+
+// An RSTNode has this specific shape
+type RSTNode = {|
+  // Either a string or a function. A string is considered a "host" node, and
+  // a function would be a composite component. It would be the component constructor or
+  // an SFC in the case of a function.
+  type: string | function
+  
+  // The props object passed to the node, which will include `children` in its raw form,
+  // exactly as it was passed to the component.
+  props: object
+  
+  // The backing instance to the node. Can be null in the case of "host" nodes and SFCs.
+  // React alternative libs can choose to provide whatever concept here applies.  If it is
+  // provided, Enzyme will expect it to follow the ComponentInterface type specified below
+  instance: ComponentInstance?
+  
+  // This is a transformed version of `props.children`. It is an array or null. If it's an array,
+  // it's an array of Nodes, meaning that they are LiteralNodes or RSTNodes. This means that 
+  // non-rendering values such as `false`, `null`, and `[]` wouldn't make it in here. This is also
+  // a "flat" array.
+  children: [Node]?
```

Comment:
If I have `<div><Foo /></div>`, and `<Foo />` renders `<Bar />`, what's going to be in `children` of `div`?

If it's truly a "transformed version of `props.children`", then it will be shallow `Foo` and which itself won't link to `Bar` RST node. Because `props.children` is shallow and does not contain rendering result.

However this doesn't make sense to me because per next comment, `rendered` is `null` for hosts. So we end up not having any way to traverse from `div` to `Bar`.

Coming back to this, it seems like `children` is not merely a "transformed version of `props.children`", but actual RST nodes corresponding to them. At which point I don't see why separate them from `rendered`. It's the same concept: an array of children nodes.

Imagine React DevTools tree. It has a universal concept of "children" (that's why it's a tree). Composites "contain" what they render, and hosts contain what's inside of them. `<div><Foo /></div>` is displayed in its fully expanded form: `<div><Foo><Bar><h1>I am bar content</h1></Bar></Foo></div>`.

I think for the same reason distinction between `rendered` and `children` is unnecessary. It is always up to traversing code "where to stop" unwrapping the tree. Composite component's children are what they render, and host component's children are based directly on what was passed to them. And you can always check if something is a composite or not by its type.